camel/data_collector/base.py

```diff
@@ -0,0 +1,159 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from typing import Any, Dict, List, Optional, Self, Tuple, Union
+
+from camel.agents.base import BaseAgent
+from camel.agents.chat_agent import ChatAgent
+from camel.messages.base import BaseMessage
+from camel.responses.agent_responses import ChatAgentResponse
+from camel.types.enums import OpenAIBackendRole
+
+
+class BaseDataCollector(ABC):
+    r"""Base class for data collectors."""
+
+    def __init__(self):
+        self.history: Dict[
+            str, List[Tuple[int, OpenAIBackendRole, BaseMessage]]
+        ] = defaultdict(list)
+        self._recording = False
+        self.agents: List[Tuple[str, BaseAgent]] = []
+        self._id = 0
+        self.data: List[Dict[str, Any]] = []
+
+    def step(
+        self,
+        message: Union[BaseMessage, ChatAgentResponse],
+        role: OpenAIBackendRole,
+        name: Optional[str] = None,
```

Comment:
these 2 information could get from `BaseMessage `

Reply:
Hi wendong, the role in BaseMessage is of type RoleType, what I need here is OpenAIBackendRole, because I need to differentiate between specific types of messages, e.g. whether it's a function call or not.